flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/reconciler/SnapshotTriggerTimestampStore.java

```diff
@@ -0,0 +1,114 @@
+     * timestamp. If none found, the return value will be the max of the resource creation timestamp
+     * and the latest triggered legacy snapshot.
+     *
+     * @param resource Flink resource
+     * @param snapshots related snapshot resources
+     * @return instant of last trigger
+     */
+    public Instant getLastPeriodicTriggerInstant(
+            AbstractFlinkResource<?, ?> resource, @Nullable Set<FlinkStateSnapshot> snapshots) {
+        if (lastTriggered.containsKey(resource.getMetadata().getUid())) {
+            return lastTriggered.get(resource.getMetadata().getUid());
+        }
+
+        return Optional.ofNullable(snapshots).orElse(Set.of()).stream()
+                .filter(s -> s.getStatus() != null && COMPLETED.equals(s.getStatus().getState()))
+                .filter(s -> (snapshotType == SAVEPOINT) == s.getSpec().isSavepoint())
+                .filter(
+                        s ->
+                                PERIODIC.name()
+                                        .equals(
+                                                s.getMetadata()
+                                                        .getLabels()
+                                                        .get(CrdConstants.LABEL_SNAPSHOT_TYPE)))
+                .map(s -> DateTimeUtils.parseKubernetes(s.getMetadata().getCreationTimestamp()))
+                .max(Comparator.naturalOrder())
+                .orElseGet(
+                        () -> {
+                            var legacyTs = getLegacyTimestamp(resource);
+                            var creationTs =
+                                    Instant.parse(resource.getMetadata().getCreationTimestamp());
+
+                            if (legacyTs.compareTo(creationTs) > 0) {
+                                return legacyTs;
+                            } else {
+                                return creationTs;
+                            }
+                        });
+    }
+
+    /**
+     * Updates the time a periodic snapshot was last triggered for this resource.
+     *
+     * @param resource Kubernetes resource
+     * @param instant new timestamp
+     */
+    public void updateLastPeriodicTriggerTimestamp(HasMetadata resource, Instant instant) {
+        lastTriggered.put(resource.getMetadata().getUid(), instant);
+    }
+
+    private Instant getLegacyTimestamp(AbstractFlinkResource<?, ?> resource) {
+        SnapshotInfo snapshotInfo;
+        switch (snapshotType) {
+            case SAVEPOINT:
+                snapshotInfo = resource.getStatus().getJobStatus().getSavepointInfo();
+                break;
+            case CHECKPOINT:
+                snapshotInfo = resource.getStatus().getJobStatus().getCheckpointInfo();
+                break;
+            default:
+                throw new IllegalArgumentException("Unsupported snapshot type: " + snapshotType);
+        }
+
+        return Instant.ofEpochMilli(snapshotInfo.getLastPeriodicTriggerTimestamp());
+    }
```

Comment:
Feels like we could get rid of the snapshotType field and pass the accessor logic in the constructor (and maybe instantiate with a static utility?)

Reply:
ah never mind I see that type field is used in other places too, we can keep it :)

Reply:
Yep, once we remove these deprecated fields, we will be able to delete that field as well.